lib/common/http1client.c

```diff
@@ -135,6 +135,9 @@ static void on_body_until_close(h2o_socket_t *sock, const char *err)
         return;
     }
 
+    client->super.bytes_read.body += sock->bytes_read;
+    client->super.bytes_read.total += sock->bytes_read;
```

Comment:
I do not think this code is correct.

`sock->bytes_read` is the total bytes that has been read from the socket. Therefore, you should not be adding that.

Instead, you should be adding delta between the each invocation of the on_xxx callbacks. Or, you can retain the value of `sock->bytes_read` before starting to receive the headers (or the body), and update `bytes_read.xxx` in each invocation by calculating the delta from the value at the start.

Also, assuming that this observation is correct, I think you might want to add a test case that slowly sends a response, so that on_body_xxx would be invoked multiple times.

Reply:
> sock->bytes_read is the total bytes that has been read from the socket

Is this true? Seeing https://github.com/h2o/h2o/blob/master/lib/common/socket/evloop.c.h#L247, we calculate `bytes_read` by taking the difference of buffer size every time we do `read(2)` and read callback is called. Doesn't it mean, it's not total bytes, and we can get the real total bytes by adding up `bytes_read`?

I added a streaming test case in https://github.com/h2o/h2o/pull/2197/commits/98bad3c105b9e3bdd26b602cc24e53e511a25386, and it passes without adding any changes to the code.

Reply:
Thank you for pointing that out. I think you are correct.

Though that raises another concern: `h2o_socket_t::bytes_written` and `h2o_socket_t::bytes_read` have different semantics, even though the names are symmetrical. The former is per-call value (as you correctly point out), whereas the latter is cumulative value.

We should align the semantics. And my preference would be to have cumulative counters for both read and write, as that would be more helpful than per-call values that can be maintained by each module while that module owns the TCP connection. Compared to that, it is much harder to maintain the amount of application data being sent / received using a socket, as the TCP socket might be passed across multiple modules.

Reply:
done in https://github.com/h2o/h2o/pull/2240/files